File: native/ios/DivinationApp/Models/CastViewModel.swift
import SwiftUI

@MainActor
final class CastViewModel: ObservableObject {
    @Published var question = ""
    @Published private(set) var snapshot: SessionSnapshot?
    @Published private(set) var lines: [CastLine] = []
    @Published private(set) var pendingLine: CastLine?
    @Published private(set) var completion: CompletionResponse?
    @Published private(set) var flowState: CastFlowState = .preparing
    @Published private(set) var isBusy = false
    @Published var errorMessage: String?
    @Published var animationToken = UUID()
    @Published var soundEnabled = UserDefaults.standard.bool(forKey: "soundEnabled") {
        didSet { UserDefaults.standard.set(soundEnabled, forKey: "soundEnabled") }
    }
    @Published var hapticsEnabled = UserDefaults.standard.object(forKey: "hapticsEnabled") as? Bool ?? true {
        didSet { UserDefaults.standard.set(hapticsEnabled, forKey: "hapticsEnabled") }
    }

    let motion = MotionShakeDetector()

    private let api = ExperienceAPI.shared
    private let vault = SessionVault()
    private let haptics = HapticEngine()
    private let sound = SoundEngine()
    private var token: String?
    private var pollingTask: Task<Void, Never>?

    init() {
        motion.onShake = { [weak self] energy in
            self?.triggerLine(energy: energy, mode: "motion")
        }
    }

    var canStart: Bool {
        !question.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !isBusy
    }

    var canTrigger: Bool {
        snapshot != nil && pendingLine == nil && lines.count < 6 && !isBusy
            && (flowState == .ready || flowState == .lineComplete)
    }

    func startSession() async {
        let trimmed = question.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        isBusy = true
        errorMessage = nil
        do {
            let created = try await api.createSession(question: trimmed)
            snapshot = created.snapshot
            lines = created.lines
            token = created.sessionToken
            vault.save(id: created.sessionId, token: created.sessionToken)
            flowState = .calibrating
            motion.start()
        } catch {
            errorMessage = error.localizedDescription
        }
        isBusy = false
    }

    func restoreSession() async {
        guard snapshot == nil, let stored = vault.load() else { return }
        isBusy = true
        do {
            let restored = try await api.getSession(id: stored.id, token: stored.token)
            snapshot = restored
            question = restored.question
            lines = restored.lines
            token = stored.token
            if restored.lineCount >= 6 {
                await fetchOrComplete(status: restored.status)
            } else {
                flowState = .calibrating
                motion.start()
            }
        } catch {
            vault.clear()
            errorMessage = "未能恢复上次会话，您可以重新开始。"
        }
        isBusy = false
    }

    func updateCalibrationState() {
        guard snapshot != nil, pendingLine == nil, lines.count < 6 else { return }
        flowState = motion.isReady ? .ready : .calibrating
    }

    func triggerLine(energy: Double, mode: String) {
        guard canTrigger else { return }
        flowState = .sensing
        haptics.validMotion(enabled: hapticsEnabled)
        isBusy = true
        Task { await lockNextLine(energy: energy, mode: mode) }
    }

    func tapFallback() {
        triggerLine(energy: max(18, motion.energy), mode: "tap")
    }

    func animationCompleted() {
        guard let pendingLine else { return }
        if !lines.contains(where: { $0.id == pendingLine.id }) {
            lines.append(pendingLine)
            lines.sort { $0.lineIndex < $1.lineIndex }
        }
        self.pendingLine = nil
        if lines.count == 6 {
            Task { await completeCast() }
        } else {
            flowState = .lineComplete
            Task {
                try? await Task.sleep(for: .milliseconds(320))
                if self.pendingLine == nil { self.flowState = .ready }
            }
        }
    }

    func reset() {
        pollingTask?.cancel()
        motion.stop()
        vault.clear()
        snapshot = nil
        token = nil
        lines = []
        pendingLine = nil
        completion = nil
        question = ""
        errorMessage = nil
        flowState = .preparing
    }

    private func lockNextLine(energy: Double, mode: String) async {
        guard let snapshot, let token else { return }
        let key = "native-\(snapshot.sessionId)-\(lines.count + 1)-\(UUID().uuidString)"
        do {
            let locked = try await api.lockNextLine(
                id: snapshot.sessionId,
                token: token,
                idempotencyKey: key,
                energy: min(100, max(0, energy)),
                triggerMode: mode
            )
            flowState = .resultLocked
            pendingLine = locked.line
            animationToken = UUID()
            flowState = .animating
            playLockedFeedback(for: locked.line)
        } catch {
            errorMessage = error.localizedDescription
            flowState = .ready
        }
        isBusy = false
    }

    private func playLockedFeedback(for line: CastLine) {
        let reducedMotion = UIAccessibility.isReduceMotionEnabled
        Task {
            let delays = reducedMotion ? [80, 150, 220] : [560, 760, 960]
            for index in 0..<3 {
                try? await Task.sleep(for: .milliseconds(delays[index] - (index == 0 ? 0 : delays[index - 1])))
                haptics.coinImpact(index: index, enabled: hapticsEnabled)
                sound.play("coin_\(index + 1)", enabled: soundEnabled)
            }
            try? await Task.sleep(for: .milliseconds(reducedMotion ? 90 : 220))
            haptics.lineLocked(moving: line.moving, enabled: hapticsEnabled)
            sound.play(line.moving ? "moving_line" : "line_lock", enabled: soundEnabled)
        }
    }

    private func completeCast() async {
        guard let snapshot, let token else { return }
        motion.stop()
        flowState = .completing
        isBusy = true
        do {
            let response = try await api.complete(id: snapshot.sessionId, token: token)
            completion = response
            lines = response.lines
            flowState = .castComplete
            haptics.castComplete(enabled: hapticsEnabled)
            sound.play("cast_complete", enabled: soundEnabled)
            if response.status == "interpretation_pending" || response.status == "interpreting" {
                startPolling()
            } else {
                vault.clear()
            }
        } catch {
            errorMessage = error.localizedDescription
            flowState = .lineComplete
        }
        isBusy = false
    }

    private func fetchOrComplete(status: String) async {
        guard let snapshot, let token else { return }
        do {
            if status == "ready_to_complete" || status == "ready" {
                completion = try await api.complete(id: snapshot.sessionId, token: token)
            } else {
                completion = try await api.result(id: snapshot.sessionId, token: token)
            }
            if let completion {
                lines = completion.lines
                flowState = .castComplete
                if completion.status == "interpretation_pending" || completion.status == "interpreting" {
                    startPolling()
                } else {
                    vault.clear()
                }
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func startPolling() {
        pollingTask?.cancel()
        guard let snapshot, let token else { return }
        pollingTask = Task {
            for _ in 0..<24 where !Task.isCancelled {
                try? await Task.sleep(for: .seconds(2))
                guard !Task.isCancelled else { return }
                do {
                    let response = try await api.result(id: snapshot.sessionId, token: token)
                    completion = response
                    if response.status == "completed" || response.status == "interpretation_failed" {
                        vault.clear()
                        return
                    }
                } catch {
                    errorMessage = "解读仍在后台生成，稍后可重新进入查看。"
                    return
                }
            }
        }
    }
}

File: native/android/app/src/main/java/com/yijing/instrument/CastViewModel.kt
package com.yijing.instrument

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID

data class CastUiState(
    val question: String = "",
    val session: SessionSnapshot? = null,
    val lines: List<CastLine> = emptyList(),
    val pendingLine: CastLine? = null,
    val completion: CompletionResponse? = null,
    val flow: CastFlowState = CastFlowState.Preparing,
    val busy: Boolean = false,
    val error: String? = null,
    val soundEnabled: Boolean = false,
    val hapticsEnabled: Boolean = true,
    val animationToken: Long = 0,
)

class CastViewModel(application: Application) : AndroidViewModel(application) {
    private val api = ExperienceApi()
    private val store = SessionStore(application)
    private val feedback = Feedback(application)
    private val preferences = application.getSharedPreferences("experience_settings", 0)
    val motion = MotionShakeDetector(application)

    private val _uiState = MutableStateFlow(
        CastUiState(
            soundEnabled = preferences.getBoolean("sound_enabled", false),
            hapticsEnabled = preferences.getBoolean("haptics_enabled", true),
        ),
    )
    val uiState: StateFlow<CastUiState> = _uiState.asStateFlow()
    private var token: String? = null

    init {
        motion.onShake = { energy -> triggerLine(energy, "motion") }
        motion.onReady = {
            _uiState.update { state ->
                if (state.session != null && state.pendingLine == null && state.lines.size < 6) {
                    state.copy(flow = CastFlowState.Ready)
                } else state
            }
        }
        restoreSession()
    }

    fun updateQuestion(value: String) {
        _uiState.update { it.copy(question = value.take(500)) }
    }

    fun startSession() {
        val question = uiState.value.question.trim()
        if (question.isEmpty() || uiState.value.busy) return
        viewModelScope.launch {
            _uiState.update { it.copy(busy = true, error = null) }
            runCatching { api.createSession(question) }
                .onSuccess { created ->
                    token = created.token
                    store.save(created.snapshot.sessionId, created.token)
                    _uiState.update {
                        it.copy(
                            session = created.snapshot,
                            lines = created.snapshot.lines,
                            flow = CastFlowState.Calibrating,
                            busy = false,
                        )
                    }
                }
                .onFailure(::showError)
        }
    }

    fun tapFallback() {
        triggerLine(uiState.value.let { motion.energy.value.coerceAtLeast(18.0) }, "tap")
    }

    fun triggerLine(energy: Double, mode: String) {
        val state = uiState.value
        if (state.session == null || state.pendingLine != null || state.lines.size >= 6 || state.busy) return
        if (state.flow != CastFlowState.Ready && state.flow != CastFlowState.LineComplete) return
        feedback.validMotion(state.hapticsEnabled)
        _uiState.update { it.copy(flow = CastFlowState.Sensing, busy = true, error = null) }
        viewModelScope.launch { lockNextLine(energy, mode) }
    }

    fun animationCompleted() {
        val pending = uiState.value.pendingLine ?: return
        val updatedLines = (uiState.value.lines + pending)
            .distinctBy { it.lineIndex }
            .sortedBy { it.lineIndex }
        _uiState.update {
            it.copy(
                lines = updatedLines,
                pendingLine = null,
                flow = if (updatedLines.size == 6) CastFlowState.Completing else CastFlowState.LineComplete,
            )
        }
        if (updatedLines.size == 6) {
            viewModelScope.launch { completeCast() }
        } else {
            viewModelScope.launch {
                delay(320)
                _uiState.update { current ->
                    if (current.pendingLine == null) current.copy(flow = CastFlowState.Ready) else current
                }
            }
        }
    }

    fun setSoundEnabled(enabled: Boolean) {
        preferences.edit().putBoolean("sound_enabled", enabled).apply()
        _uiState.update { it.copy(soundEnabled = enabled) }
    }

    fun setHapticsEnabled(enabled: Boolean) {
        preferences.edit().putBoolean("haptics_enabled", enabled).apply()
        _uiState.update { it.copy(hapticsEnabled = enabled) }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    fun reset() {
        motion.stop()
        store.clear()
        token = null
        _uiState.update {
            CastUiState(
                soundEnabled = it.soundEnabled,
                hapticsEnabled = it.hapticsEnabled,
            )
        }
    }

    private fun restoreSession() {
        val stored = store.load() ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(busy = true) }
            runCatching { api.getSession(stored.id, stored.token) }
                .onSuccess { restored ->
                    token = stored.token
                    _uiState.update {
                        it.copy(
                            question = restored.question,
                            session = restored,
                            lines = restored.lines,
                            flow = if (restored.lines.size < 6) CastFlowState.Calibrating else CastFlowState.Completing,
                            busy = false,
                        )
                    }
                    if (restored.lines.size >= 6) fetchOrComplete(restored.status)
                }
                .onFailure {
                    store.clear()
                    _uiState.update { state ->
                        state.copy(busy = false, error = "未能恢复上次会话，您可以重新开始。")
                    }
                }
        }
    }

    private suspend fun lockNextLine(energy: Double, mode: String) {
        val session = uiState.value.session ?: return
        val activeToken = token ?: return
        val key = "native-${session.sessionId}-${uiState.value.lines.size + 1}-${UUID.randomUUID()}"
        runCatching {
            api.lockNextLine(session.sessionId, activeToken, key, energy, mode)
        }.onSuccess { locked ->
            _uiState.update {
                it.copy(
                    pendingLine = locked.line,
                    flow = CastFlowState.ResultLocked,
                    busy = false,
                    animationToken = System.nanoTime(),
                )
            }
            _uiState.update { it.copy(flow = CastFlowState.Animating) }
            playLockedFeedback(locked.line)
        }.onFailure {
            _uiState.update { state ->
                state.copy(error = it.message ?: "起爻失败，请重试。", flow = CastFlowState.Ready, busy = false)
            }
        }
    }

    private fun playLockedFeedback(line: CastLine) {
        viewModelScope.launch {
            val state = uiState.value
            delay(560)
            feedback.coinImpact(0, state.hapticsEnabled, state.soundEnabled)
            delay(200)
            feedback.coinImpact(1, state.hapticsEnabled, state.soundEnabled)
            delay(200)
            feedback.coinImpact(2, state.hapticsEnabled, state.soundEnabled)
            delay(220)
            feedback.lineLocked(line.moving, state.hapticsEnabled, state.soundEnabled)
        }
    }

    private suspend fun completeCast() {
        val session = uiState.value.session ?: return
        val activeToken = token ?: return
        motion.stop()
        _uiState.update { it.copy(flow = CastFlowState.Completing, busy = true) }
        runCatching { api.complete(session.sessionId, activeToken) }
            .onSuccess { completed ->
                _uiState.update {
                    it.copy(
                        completion = completed,
                        lines = completed.lines,
                        flow = CastFlowState.CastComplete,
                        busy = false,
                    )
                }
                val state = uiState.value
                feedback.castComplete(state.hapticsEnabled, state.soundEnabled)
                if (completed.status in pendingStatuses) pollResult() else store.clear()
            }
            .onFailure(::showError)
    }

    private suspend fun fetchOrComplete(status: String) {
        val session = uiState.value.session ?: return
        val activeToken = token ?: return
        runCatching {
            if (status == "ready" || status == "ready_to_complete") {
                api.complete(session.sessionId, activeToken)
            } else {
                api.result(session.sessionId, activeToken)
            }
        }.onSuccess { completed ->
            _uiState.update {
                it.copy(
                    completion = completed,
                    lines = completed.lines,
                    flow = CastFlowState.CastComplete,
                    busy = false,
                )
            }
            if (completed.status in pendingStatuses) pollResult() else store.clear()
        }.onFailure(::showError)
    }

    private fun pollResult() {
        val session = uiState.value.session ?: return
        val activeToken = token ?: return
        viewModelScope.launch {
            repeat(24) {
                delay(2_000)
                val completed = runCatching { api.result(session.sessionId, activeToken) }
                    .getOrElse {
                        _uiState.update { state ->
                            state.copy(error = "解读仍在后台生成，稍后可重新进入查看。")
                        }
                        return@launch
                    }
                _uiState.update { it.copy(completion = completed, lines = completed.lines) }
                if (completed.status !in pendingStatuses) {
                    store.clear()
                    return@launch
                }
            }
        }
    }

    private fun showError(error: Throwable) {
        _uiState.update {
            it.copy(
                busy = false,
                error = error.message ?: "服务暂时不可用，请稍后重试。",
            )
        }
    }

    override fun onCleared() {
        motion.stop()
        feedback.release()
        super.onCleared()
    }

    private companion object {
        val pendingStatuses = setOf("interpretation_pending", "interpreting")
    }
}
